### PR TITLE
[ao] making hist_obs handle torch.inf and closeby values

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103467

Summary: This PR does 2 things:

1) Previously this would simply error, now it will ignore any
torch.inf values that it recieves. note: The code checks for torch.inf after
aminmax that way if there are no torch.inf values found, the perf is a
relatively unchanged

2) as mentioned in https://github.com/pytorch/pytorch/issues/100051,
values close to (but not quite at) the maximum/minimum float value could
overflow to infinity in the course of _adjust_min_max() (when this large
value would be multiplied by something in the middle of a calculation
that would otherwise result in a non inf value). This was fixed by
rearranging the order of operations for the lines in question without
altering the actual equations. Specifically, where operations in lines
1095, 1098 and 1100 have multiplication and division of large values,
its better to divide the two large values before multiplying, rather
than multiplying the two large values together (creating overflow) before dividing like it had been.

Test Plan: python test/test_quantization.py
TestObserver.test_histogram_observer_ignore_infinity

python test/test_quantization.py TestObserver.test_histogram_observer_handle_close_to_infinity
Reviewers:

Subscribers:

Tasks:

Tags:

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Differential Revision: [D51489345](https://our.internmc.facebook.com/intern/diff/D51489345)

### CODE REVIEW

File: torch/ao/quantization/observer.py [hunk 1]
```diff
@@ -1143,12 +1143,21 @@ def forward(self, x_orig: torch.Tensor) -> torch.Tensor:
         if x_orig.numel() == 0:
             return x_orig
         x = x_orig.detach()
+        x_min, x_max = torch.aminmax(x)
+        # want to ignore torch.inf since we don't actually
+        # want to make our quantization range infinite
+        # and in practice those values will be clamped
+        if x_min==-torch.inf or x_max==torch.inf:
```
Comment (leslie-fang-intel) [inline torch/ao/quantization/observer.py:1150]: Thanks for the fix. Looks like the example in https://github.com/pytorch/pytorch/issues/100051 still fails. It may have case that `-3.4028234663852886 * 10**38` is not `-torch.inf`, but after it multiply by `self.upsample_rate` here https://github.com/pytorch/pytorch/blob/4c3799447f4cb8157e2145446c659b8629cb3c08/torch/ao/quantization/observer.py#L1095, it will change to `-torch.inf`.

**Comment by leslie-fang-intel** [comment]
Hi @HDCharles, thanks for the fix.

I think this fix still has accuracy issue for some cases.
For example in LLM models, there is a pattern of 
https://github.com/huggingface/transformers/blob/70c79940957fb25b54bd1b106935c756b90345eb/src/transformers/models/gptj/modeling_gptj.py#L171-L175.

```
attn_weights     attention_mask
      \                       /
              add
                |
            softmax
```

The `attention_mask` is a tensor with values of `0` or `-3.4028234663852886 * 10**38` in order to filter out some `attn_weights` values with `softmax`. So if we insert `HistogramObserver` of `obs2` as

```
attn_weights     attention_mask
      \                       /
      obs1             obs2
          \                /
              add
                |
            softmax
```

With current fix, the `HistogramObserver` will still record the input values of 0. Then convert it to quant/dequant, the scale is `1.1921e-07` and zp is `0` as I got from below example code. So it will cause wrong result of `attention_mask -> quant -> dequant->softmax` after convert comparing with original `attention_mask -> softmax`. 

@jianan-gu please to check my description for LLM use case.

I also create a example code as:
```
    obser = HistogramObserver.with_args(reduce_range=False)()
    mask = torch.tensor([-3.4028234663852886 * 10**38, -0.0])
    obser(mask)
    obser(mask)
    scale, zp = obser.calculate_qparams()
    print("scale is: {}".format(scale))
    print("zp is: {}".format(zp))

    input = torch.tensor([2.1, 3.2])
    ref_result = torch.softmax(input + mask, dim=0)
    print("ref_result is: {}".format(ref_result))

    quant_tensor = torch.quantize_per_tensor(mask, scale, zp, torch.quint8)
    dequant_tensor = quant_tensor.dequantize()
    result = torch.softmax(input + dequant_tensor, dim=0)
    print("result is: {}".format(result))
```

specific for this issue, since the overflow happens at `_adjust_min_max` but the original tensor didn't overflow, may be we can consider switch the calculation order inside  `_adjust_min_max` to avoid overflow?

**Comment by HDCharles** [comment]
> 
> ```
> attn_weights     attention_mask
>       \                       /
>       obs1             obs2
>           \                /
>               add
>                 |
>             softmax
> ```
> 

@leslie-fang-intel are you trying to get a float out of the add operation? or are you trying to carry the quantized computation the whole way through?

**Comment by leslie-fang-intel** [comment]
> > ```
> > attn_weights     attention_mask
> >       \                       /
> >       obs1             obs2
> >           \                /
> >               add
> >                 |
> >             softmax
> > ```
> 
> @leslie-fang-intel are you trying to get a float out of the add operation? or are you trying to carry the quantized computation the whole way through?

Oh, sorry for the confusion. Is the output dtype (float or uint8) of `add` makes any difference to this issue?

File: torch/ao/quantization/observer.py [hunk 2]
```diff
@@ -1092,10 +1092,10 @@ def _adjust_min_max(
...
         downsample_rate = int(
             torch.ceil(
-                (combined_max - combined_min) * upsample_rate / (self.max_val - self.min_val)
+                ((combined_max - combined_min) / (self.max_val - self.min_val)) * upsample_rate
             ).item()
         )
-        e = downsample_rate * (self.max_val - self.min_val) / upsample_rate - (combined_max - combined_min)
+        e = downsample_rate / upsample_rate * (self.max_val - self.min_val) - (combined_max - combined_min)
         start_idx = int(
             torch.round((self.min_val - combined_min) * self.bins * upsample_rate / (self.max_val - self.min_val)).item()
```
Comment (leslie-fang-intel) [inline torch/ao/quantization/observer.py:1100]: Suggest to also adjust the order of calculation of `start_idx`.

**Comment by HDCharles** [comment]
> > > ```
> > > attn_weights     attention_mask
> > >       \                       /
> > >       obs1             obs2
> > >           \                /
> > >               add
> > >                 |
> > >             softmax
> > > ```
> > 
> > 
> > @leslie-fang-intel are you trying to get a float out of the add operation? or are you trying to carry the quantized computation the whole way through?
> 
> Oh, sorry for the confusion. Is the output dtype (float or uint8) of `add` makes any difference to this issue?

if you want a quantized tensor that can both reach to infinity and represent the normal values from attn_weights (i.e. the result after add) i don't think thats going to be possible without some finagling and the whole approach may need to be rethought regardless of the PR

**Comment by leslie-fang-intel** [comment]
> > > > ```
> > > > attn_weights     attention_mask
> > > >       \                       /
> > > >       obs1             obs2
> > > >           \                /
> > > >               add
> > > >                 |
> > > >             softmax
> > > > ```
> > > 
> > > 
> > > @leslie-fang-intel are you trying to get a float out of the add operation? or are you trying to carry the quantized computation the whole way through?
> > 
> > 
> > Oh, sorry for the confusion. Is the output dtype (float or uint8) of `add` makes any difference to this issue?
> 
> if you want a quantized tensor that can both reach to infinity and represent the normal values from attn_weights (i.e. the result after add) i don't think thats going to be possible without some finagling and the whole approach may need to be rethought regardless of the PR

Hi @HDCharles, thanks for the comment. So the general solutions in this PR is:
1. Filter out the input values of  infinity at input of the observer.
2. Adjust the computation order inside observer to avoid the case: the values is not infinity at input but may become infinity due the calculation inside observer.
I think it should work for my use case.

File: test/quantization/core/test_workflow_module.py [hunk 3]
```diff
@@ -327,6 +327,35 @@ def test_histogram_observer_consistent_buffer_shape(self):
...
+        """
+        Ensures that HistogramObserver doesn't record values of infinity
+        """
+        obs = HistogramObserver()
+        obs2 = HistogramObserver()
+        x = torch.randn(4, 4, 4, 4)
+        obs(x * torch.inf)
+        obs(x)
+        obs2(x)
+        obs(x * torch.inf)
```
Comment (jcaip) [inline test/quantization/core/test_workflow_module.py:340]: Is this meant to check negative infinity?

File: torch/ao/quantization/observer.py [hunk 4]
```diff
@@ -1092,12 +1092,12 @@ def _adjust_min_max(
         # Therefore, we avoid such division operation.
         downsample_rate = int(
             torch.ceil(
-                (combined_max - combined_min) * upsample_rate / (self.max_val - self.min_val)
+                ((combined_max - combined_min) / (self.max_val - self.min_val)) * upsample_rate
             ).item()
         )
-        e = downsample_rate * (self.max_val - self.min_val) / upsample_rate - (combined_max - combined_min)
+        e = downsample_rate / upsample_rate * (self.max_val - self.min_val) - (combined_max - combined_min)
```
Comment (jcaip) [inline torch/ao/quantization/observer.py:1098]: Are downsample_rate and upsample_rate ever close to each other? 

If they're equal I think this will be resolve to (1) * (something close to infinity) which I think still might overflow?

File: test/quantization/core/test_workflow_module.py (hunk 3 above)
Comment (jcaip) [inline test/quantization/core/test_workflow_module.py:340]: Is this meant to check negative infinity?

**Review by jcaip — COMMENTED** [review]
couple comments, otherwise looks good

File: test/quantization/core/test_workflow_module.py (hunk 3 above)
Comment (HDCharles) [reply test/quantization/core/test_workflow_module.py:340]: yeah it checks both positive and negative infinity, basically running obs(x*torch.inf) shouldn't do anything.

File: torch/ao/quantization/observer.py (hunk 4 above)
Comment (HDCharles) [reply torch/ao/quantization/observer.py:1098]: that isn't an issue, 1*(something representable by float) won't overflow

File: torch/ao/quantization/observer.py [hunk 5]
```diff
@@ -1092,15 +1092,17 @@ def _adjust_min_max(
         # Therefore, we avoid such division operation.
         downsample_rate = int(
```
Comment (jerryzh168) [inline torch/ao/quantization/observer.py:1093]: can you add some docs on the order of operations, when does this order work or does not work etc.?

File: torch/ao/quantization/observer.py [hunk 6]
```diff
@@ -1092,15 +1092,17 @@ def _adjust_min_max(
...
-        e = downsample_rate * (self.max_val - self.min_val) / upsample_rate - (combined_max - combined_min)
+        e = downsample_rate / upsample_rate * (self.max_val - self.min_val) - (combined_max - combined_min)
         start_idx = int(
-            torch.round((self.min_val - combined_min) * self.bins * upsample_rate / (self.max_val - self.min_val)).item()
+            torch.round((self.min_val - combined_min) / (self.max_val - self.min_val) * self.bins * upsample_rate).item()
         )
         combined_max = combined_max + e
         combined_min = combined_min
+        assert not torch.isinf(torch.tensor(combined_max - combined_min)), "combined min/max are inf"
+        assert not torch.isinf(torch.tensor(downsample_rate)), "downsample"
```
Comment (jerryzh168) [inline torch/ao/quantization/observer.py:1105]: nit: seems error message is not complete

File: torch/ao/quantization/observer.py [hunk 7]
```diff
@@ -1143,12 +1145,21 @@ def forward(self, x_orig: torch.Tensor) -> torch.Tensor:
...
+            if x.numel() == 0:
+                return x_orig
+            x_min, x_max = torch.aminmax(x)
         min_val = self.min_val
         max_val = self.max_val
         same_values = min_val.item() == max_val.item()
         is_uninitialized = min_val == float("inf") and max_val == float("-inf")
         if is_uninitialized or same_values:
-            min_val, max_val = torch.aminmax(x)
+            min_val, max_val = x_min , x_max
```
Comment (jerryzh168) [inline torch/ao/quantization/observer.py:1162]: nit: why there is a space after x_min?

File: torch/ao/quantization/observer.py (hunk 6 above)
Comment (HDCharles) [reply torch/ao/quantization/observer.py:1105]: yeah the PR is not passing CI and I don't know why. The tests pass on my machine but not in CI. Any advice would be appreciated, its the new 'test close to infinity' test that's failing too.

**Comment by andrewor14** [comment]
Hey @HDCharles @jerryzh168 what is the status of this PR? I feel we are close to landing. Should we rebase and try to land?

**Comment by jerryzh168** [comment]
is histogram really defined for torch.inf?

File: torch/ao/quantization/observer.py [hunk 8]
```diff
@@ -1143,12 +1142,21 @@ def forward(self, x_orig: torch.Tensor) -> torch.Tensor:
         if x_orig.numel() == 0:
             return x_orig
         x = x_orig.detach()
+        x_min, x_max = torch.aminmax(x)
+        # want to ignore torch.inf since we don't actually
```
Comment (jerryzh168) [inline torch/ao/quantization/observer.py:1146]: oh OK so it's ignored.. that seems reasonable

File: torch/ao/quantization/observer.py [hunk 9]
```diff
@@ -1092,15 +1092,14 @@ def _adjust_min_max(
         # Therefore, we avoid such division operation.
         downsample_rate = int(
             torch.ceil(
-                (combined_max - combined_min) * upsample_rate / (self.max_val - self.min_val)
+                ((combined_max - combined_min) / (self.max_val - self.min_val)) * upsample_rate
```
Comment (jerryzh168) [inline torch/ao/quantization/observer.py:1095]: I'm not sure if changing these orders will cause any issues for existing tests, might be good to import to fbcode and make sure CI is green

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51489345).

**Review by andrewor14 — APPROVED** [review]

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51489345).

**Comment by HDCharles** [comment]
pytorchmergebot merge
